lib/render/ContourRenderer.cpp

```diff
@@ -111,10 +145,57 @@ int ContourRenderer::_buildCache()
     Grid::CopyToArr3(_cacheParams.boxMax, boxMax);
 
     Grid *grid = _dataMgr->GetVariable(_cacheParams.ts, _cacheParams.varName, _cacheParams.level, _cacheParams.lod, boxMin, boxMax);
-    Grid *heightGrid = NULL;
-    if (!_cacheParams.heightVarName.empty()) { heightGrid = _dataMgr->GetVariable(_cacheParams.ts, _cacheParams.heightVarName, _cacheParams.level, _cacheParams.lod, boxMin, boxMax); }
+    Grid *grid2 = nullptr;
+    Grid *heightGrid = nullptr;
+
+    int dims = grid->GetGeometryDim();
+    if (!_cacheParams.heightVarName.empty() && dims == 2) { heightGrid = _dataMgr->GetVariable(_cacheParams.ts, _cacheParams.heightVarName, _cacheParams.level, _cacheParams.lod, boxMin, boxMax); }
+    if (grid == NULL || (heightGrid == NULL && !_cacheParams.heightVarName.empty())) {
+        if (grid) delete grid;
+        if (grid2) delete grid2;
+        if (heightGrid) delete heightGrid;
+        return -1;
+    }
 
-    if (grid == NULL || (heightGrid == NULL && !_cacheParams.heightVarName.empty())) { return -1; }
+    if (dims == 3) {
+        planeDescription pd;
+        pd.boxMin = ToCoordType(_cacheParams.boxMin);
+        pd.boxMax = ToCoordType(_cacheParams.boxMax);
+        pd.rotation = _cacheParams.sliceRotation;
+        pd.origin = _cacheParams.sliceOrigin;
+        pd.sideSize = _cacheParams.sliceResolution;
+
+        auto o = ToVec3(_cacheParams.sliceOrigin);
+        auto n = ArbitrarilyOrientedRegularGrid::GetOffsetNormal(pd);
+        auto offsetOrigin = o + n * (float)_cacheParams.sliceOffset;
+        pd.origin = ToDoubleVec(offsetOrigin);
+
+        DimsType dims = {pd.sideSize, pd.sideSize, 1};
+
+        ArbitrarilyOrientedRegularGrid *grid2d = new ArbitrarilyOrientedRegularGrid(grid, pd, dims);
+        grid2 = grid;
+        grid = grid2d;
+
+        CoordType corner1, corner2, corner3, corner4;
+        grid2d->GetUserCoordinates({0, 0, 0}, corner1);
+        grid2d->GetUserCoordinates({0, pd.sideSize - 1, 0}, corner2);
+        grid2d->GetUserCoordinates({pd.sideSize - 1, 0, 0}, corner3);
+        grid2d->GetUserCoordinates({pd.sideSize - 1, pd.sideSize - 1, 0}, corner4);
+
+        _sliceQuad = {
+            glm::vec3(corner1[0], corner1[1], corner1[2]),
+            glm::vec3(corner3[0], corner3[1], corner3[2]),
+            glm::vec3(corner4[0], corner4[1], corner4[2]),
+            glm::vec3(corner2[0], corner2[1], corner2[2]),
+        };
```

Comment:
The ordering for these vertices is also defined in SliceRenderer.cpp:217.  Are we sure this ordering (corner1, 3, 4, then 2) shouldn't be kept in one place, the ArbitrarilyOrientedRegularGrid?

Reply:
I believe this code used to be in the grid class but it was moved to the slice renderer in the PR review? They are based on the grid being CCW which should always be consistent anyway.

Reply:
@sgpearse the Grid class uses CCW ordering. So I'm not sure what you are referring to when you say 'this info should be retrieved to prevent duplication"

Reply:
My bad - I was confusing this with something else.